File: rusty_web/src/api/jobs.rs
use commons::errors::RustyError;
use domain::jobs::{Job, RegisterJob};

use crate::api::client::reqwasm_post;
use crate::api::utils::parse_entries;

/// Function to retrieve jobs for project from a GraphQL endpoint.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn get_jobs_for_project(project_id: String) -> Result<Vec<Job>, RustyError> {
    let payload = serde_json::json!({
        "query": format!(r#"query {{
            jobs {{
                get(filter: {{
                    project_id: "{}"
                }}) {{
                    id
                    name
                    description
                    template
                    projectId
                }}
            }}
        }}"#, project_id),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: serde_json::Value = serde_json::from_str(&data)?;
    let json_data = json_data["data"]["jobs"]["get"].clone();
    parse_entries(json_data)
}

/// Function to retrieve a job from a GraphQL endpoint by id.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn get_job(id: String) -> Result<Job, RustyError> {
    let payload = serde_json::json!({
        "query": format!(r#"query {{
            jobs {{
                getById(id: "{}") {{
                    id
                    name
                    description
                    template
                    projectId
                }}
            }}
        }}"#, id),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: serde_json::Value = serde_json::from_str(&data)?;
    let json_data = json_data["data"]["jobs"]["getById"].clone();
    parse_entries(json_data)
}

/// Function to register a new job via GraphQL endpoint.
///
/// # Errors
///
/// This function can generate the following errors:
///
/// * `RustyError` - If there was an error during the creation of the item.
#[allow(clippy::future_not_send)]
pub async fn register_job(model: RegisterJob) -> Result<String, RustyError> {
    let description = model
        .description
        .map_or_else(String::new, |desc| format!("description: \"{desc}\""));
    let payload = serde_json::json!({
        "query": format!(r#"mutation {{
            jobs {{
                register(job: {{
                    name: "{}"
                    {}
                    template: "{}"
                    projectId: "{}"
                }})
            }}
        }}"#, model.name, description, model.template, model.project_id),
        "variables": {}
    });

    let data = reqwasm_post(&payload).await?;
    let json_data: serde_json::Value = serde_json::from_str(&data)?;
    Ok(json_data["data"]["jobs"]["register"].to_string())
}

File: rusty_web/src/lib/api/jobs.ts
import { fetchPost } from '$lib/utils/api';
import type { Job } from '$lib/domain/job';

const getProjectJobsQuery = (page: number, id: string, name: string) => {
	const filter = `filter: { project_id: { equals: "${id}" }, name: { contains: "${name}" } }, `;
	const options = `options: { pageNumber: ${page}, pageSize: 50, sortMode: ASCENDING, sortField: "name" }`;

	return `query {
		jobs {
			get(${filter}${options}){
				total
				page
				pageSize
				entries {
					id
					name
					description
					projectId
					pipelines {
						id
						number
						status
						stageStatus
						registerDate
					}
				}
			}
		}
	}`;
};

export const getProjectJobs = async (auth: string, page: number, id: string, name: string) => {
	try {
		const response = await fetchPost(
			auth,
			JSON.stringify({ query: getProjectJobsQuery(page, id, name) })
		);

		if (!response.ok) {
			return {
				errors: ['Fetch project jobs failed']
			};
		} else {
			const { data, errors } = await response.json();
			if (errors && errors.length > 0) {
				return {
					errors: errors.map((error: { message: string }) => error.message)
				};
			} else if (data) {
				const paged = data?.jobs?.get;
				const jobs: Job[] = paged?.entries ?? [];
				return {
					total: paged?.total ?? 0,
					page: paged?.page ?? 1,
					pageSize: paged?.pageSize ?? 20,
					entries: jobs
				};
			}
		}
	} catch (error) {
		return {
			errors: ['Fetch project jobs failed']
		};
	}
};

const getJobByIdQuery = (id: string) => {
	return `query {
		jobs {
			getById(id: "${id}"){
				id
				name
				description
				template
			}
		}
	}`;
};

export const getJobById = async (auth: string, id: string) => {
	try {
		const response = await fetchPost(auth, JSON.stringify({ query: getJobByIdQuery(id) }));

		if (!response.ok) {
			return {
				errors: ['Get job by id failed']
			};
		} else {
			const { data, errors } = await response.json();
			if (errors && errors.length > 0) {
				return {
					errors: errors.map((error: { message: string }) => error.message)
				};
			} else if (data) {
				return data?.jobs?.getById;
			}
		}
	} catch (error) {
		return {
			errors: ['Get job by id failed']
		};
	}
};
